### ISSUE TITLE
list index out of range for power_avg

### ISSUE THREAD

**Issue by agarwalmanvi** [lfwa/carbontracker#29]
Thanks for the nice tool! I'm trying to get this to work with [GeNN](http://genn-team.github.io/genn/) but I get the following error:
```
Traceback (most recent call last):
  File "/home/manvi/Documents/carbontracker/carbontracker/tracker.py", line 281, in epoch_end
    self.tracker.epoch_end()
  File "/home/manvi/Documents/carbontracker/carbontracker/tracker.py", line 139, in epoch_end
    self._log_epoch_measurements()
  File "/home/manvi/Documents/carbontracker/carbontracker/tracker.py", line 162, in _log_epoch_measurements
    power_avg = np.mean(comp.power_usages[-1], axis=0)
IndexError: list index out of range
```
To make sure the error is not caused by GeNN, I tried to use a tracker with some plain Python code, and it still gave me the same error. I tried to identify where the error is coming from within the source code of carbontracker, but wasn't successful. Any ideas on how to fix this? I'm using Ubuntu with Intel i3.

**Comment by lfwa** [comment]
Hi,

We appreciate the feedback. :)

Could you provide the plain Python code that gave you the error as well as some more details of the environment that you are trying to run it in (e.g. Python version, Ubuntu version, and CPU generation)?

**Comment by agarwalmanvi** [comment]
Thanks for the prompt response! Here's the plain Python code:
```
from carbontracker.tracker import CarbonTracker

TRIALS = 100
tracker = CarbonTracker(epochs=TRIALS)

for trial in range(TRIALS):

    tracker.epoch_start()
    print("Trial: " + str(trial))
    sum = 1200 + 100
    tracker.epoch_end()

tracker.stop()
```
I'm running it in a conda environment with Python 3.7.7. My Ubuntu OS is version 18.04 and my CPU is from family 6 model 42.

**Comment by lfwa** [comment]
Your epoch durations may be too short for **carbontracker** to be able to collect any measurements. To verify this, could you try replacing the addition  `sum = 1200 + 100` with `time.sleep(5)` and decreasing the `update_interval` argument to 1:

```
from carbontracker.tracker import CarbonTracker
import time

TRIALS = 100
tracker = CarbonTracker(epochs=TRIALS, update_interval=1)

for trial in range(TRIALS):

    tracker.epoch_start()
    print("Trial: " + str(trial))
    time.sleep(5)
    tracker.epoch_end()

tracker.stop()
```
If this is the case, I will look into adding a more appropriate error message.

**Comment by agarwalmanvi** [comment]
That worked but now I get a different error (also to do with out of range index):
```
Traceback (most recent call last):
  File "/home/manvi/Documents/carbontracker/carbontracker/tracker.py", line 284, in epoch_end
    self._output_actual()
  File "/home/manvi/Documents/carbontracker/carbontracker/tracker.py", line 358, in _output_actual
    _co2eq = self._co2eq(energy)
  File "/home/manvi/Documents/carbontracker/carbontracker/tracker.py", line 381, in _co2eq
    ci = self.intensity_updater.average_carbon_intensity(pred_time_dur)
  File "/home/manvi/Documents/carbontracker/carbontracker/tracker.py", line 55, in average_carbon_intensity
    ci.carbon_intensity = self.carbon_intensities[-1].carbon_intensity
IndexError: list index out of range
```
Another question: the computations I'm doing with GeNN are very short. For example, I printed the duration from [this line](https://github.com/lfwa/carbontracker/blob/1c9307b5fc2a408667f3a19c12c2b45be08354b2/carbontracker/tracker.py#L153) for the first epoch and it came to 0:00:00.89. This was probably why I was getting that error with my GeNN script. Is reducing the value of `update_interval` to a small enough value useful (0.1 for example?) in this case?

**Comment by lfwa** [comment]
Hi again,

I was unfortunately not able to replicate your error. I will create a separate issue for it and take a look at it in the very near future. If you can test it with a reduced `update_interval` on your GeNN script that would be perfect. :)

Regarding your second question:
Yes, you should be able to reduce `update_interval` to 0 or a very small value (the argument specifies the time *slept* between measurements). If it is not able to gather a measurement for every epoch, it should extrapolate the latest measurement to the previous unmeasured epochs. This, however, assumes that you entire training session is long enough for any measurements to be gathered at all, i.e., if you training session is very short (a few seconds) and it is not able to gather a measurement in this period then it may fail.

**Comment by agarwalmanvi** [comment]
Thanks, it would be great to have a working example to understand how carbontracker can be used!

I tried my GeNN script again with `update_interval` set to 0 and 0.0001. Both gave the first error (the one from `_log_epoch_measurements`). The entire training session for the GeNN script I'm referring to (without any extra stuff like plotting or printing information) should take tens of minutes on my hardware since I'm doing thousands of trials.

**Comment by lfwa** [comment]
Okay, that is unfortunate. I will look into the issues with very short epoch durations, but it may take a while before we can release a new version with a fix.

If you are not interested in the predictions but just want to monitor your carbon footprint, you can treat your entire script as a single epoch, similar to this:
```
from carbontracker.tracker import CarbonTracker

tracker = CarbonTracker(epochs=1, epochs_before_pred=0)
tracker.epoch_start()

# Your script.

tracker.epoch_end()
tracker.stop()
```

We will be updating the README to be more comprehensive soon together with a potential release of version 2.0 making **carbontracker** easier to use and setup.

**Comment by agarwalmanvi** [comment]
Indeed my first goal was to simply track the carbon footprint. Thanks for your suggestion on treating the whole script like one epoch. It was able to move past the error with `_log_epoch_measurements` but hit the same error with `carbon_intensity` as described above. I will track [issue #30](https://github.com/lfwa/carbontracker/issues/30) for a potential resolution.

**Comment by Princec711** [comment]
Issue 1 : Actually I tried above test code in jupyter notebook and getting below error.

Error : 

Trial: 0
CarbonTracker: The following components were found: GPU with device(s) GeForce GTX 1050 Ti.
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py", line 274, in epoch_end
    self.tracker.epoch_end()
  File "C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py", line 139, in epoch_end
    self._log_epoch_measurements()
  File "C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py", line 163, in _log_epoch_measurements
    power_avg = np.mean(comp.power_usages[-2], axis=0)
IndexError: list index out of range

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py in epoch_end(self)
    273         try:
--> 274             self.tracker.epoch_end()
    275 

C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py in epoch_end(self)
    138         self.epoch_times.append(time.time() - self.cur_epoch_time)
--> 139         self._log_epoch_measurements()
    140 

C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py in _log_epoch_measurements(self)
    162             if np.isnan(power_avg).all():
--> 163                 power_avg = np.mean(comp.power_usages[-2], axis=0)
    164             self.logger.info(

IndexError: list index out of range

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
<ipython-input-1-a721ced6caeb> in <module>
     10     print("Trial: " + str(trial))
     11     time.sleep(5)
---> 12     tracker.epoch_end()
     13 
     14 tracker.stop()

C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py in epoch_end(self)
    285                 self._delete()
    286         except Exception as e:
--> 287             self._handle_error(e)
    288 
    289     def stop(self):

C:\ProgramData\Anaconda3\lib\site-packages\carbontracker\tracker.py in _handle_error(self, error)
    326             self._delete()
    327         else:
--> 328             sys.exit(os.EX_SOFTWARE)
    329 
    330     def _output_energy(self, description, time, energy, co2eq, conversions):

AttributeError: module 'os' has no attribute 'EX_SOFTWARE'
